src/ForwardPipe.h

```diff
@@ -49,6 +49,22 @@ class ForwardPipe {
                                const std::vector<float>& means_2,
                                const std::vector<float>& variances_2) = 0;
 
+    // have a better way to deal with situations when caller doesn't need the prior net?
+    // implement this.
+    virtual void move_residual(unsigned int filter_size,
```

Comment:
Can't you support both cases by the combination of emplace_back and the universal references?